wdl/src/main/scala/wdl/WdlGraphNode.scala

```diff
@@ -99,21 +102,30 @@ object WdlGraphNode {
       // - The outputs from any new input Nodes we had to make (so that other Nodes don't have to recreate them)
       FoldState(
         nodes = acc.nodes + gnani.node ++ gnani.newInputs ++ usedOuterGraphInputNodes ++ gnani.newExpressions,
-        availableInputs = acc.availableInputs ++ newCallOutputPorts ++ newInputOutputPorts
+        availableInputs = acc.availableInputs ++ newCallOutputPorts ++ newInputOutputPorts ++ newOginOutputs
       )
     }
 
     def buildNode(acc: FoldState, node: WdlGraphNode): ErrorOr[FoldState] = node match {
-      case wdlCall: WdlCall => WdlCall.buildWomNodeAndInputs(wdlCall, acc.availableInputs, outerLookup, preserveIndexForOuterLookups) map { case cnani @ CallNodeAndNewNodes(call, _, _) =>
+      case wdlCall: WdlCall => WdlCall.buildWomNodeAndInputs(wdlCall, acc.availableInputs, outerLookup, preserveIndexForOuterLookups) map { case cnani @ CallNodeAndNewNodes(call, _, _, _) =>
         foldInGeneratedNodeAndNewInputs(acc, call.localName + ".")(cnani)
       }
 
-      case decl: DeclarationInterface => Declaration.buildWomNode(decl, acc.availableInputs, outerLookup, preserveIndexForOuterLookups) map { declNode =>
+      case decl: DeclarationInterface => Declaration.buildWomNode(decl, acc.availableInputs, outerLookup, preserveIndexForOuterLookups) map { declNodeMaker =>
         // As with GeneratedNodeAndNewInputs, we might have made some new OuterGraphInputNodes to build this DeclarationNode, so
         // make sure they get included:
-        val newOgins: Set[OuterGraphInputNode] = declNode.toGraphNode.upstream.filterByType[OuterGraphInputNode]
-        val newOginOutputs = newOgins map {ogin => ogin.localName -> ogin.singleOutputPort }
-        FoldState(acc.nodes + declNode.toGraphNode ++ newOgins, acc.availableInputs ++ newOginOutputs ++ declNode.singleOutputPort.map { sop => declNode.toGraphNode.localName -> sop })
+        val declNode = declNodeMaker.toGraphNode
+        val newOgins: Set[OuterGraphInputNode] = declNode.upstreamAncestry.filterByType[OuterGraphInputNode]
+        val newOginOutputs = newOgins map { ogin => ogin.localName -> ogin.singleOutputPort }
+
+        // We only want to include the output port as a input to other GraphNodes in the FoldState if it's not a Workflow Output:
+        val availableOutputPort: Option[(String, OutputPort)] = declNodeMaker match {
+          case InputDeclarationNode(graphInputNode) => Option(declNode.localName -> graphInputNode.singleOutputPort)
+          case IntermediateValueDeclarationNode(expressionNode) => Option(declNode.localName -> expressionNode.singleExpressionOutputPort)
+          case GraphOutputDeclarationNode(_) => None
```

Comment:
You'll notice that I originally thought exactly the same, and then my second commit had to undo that.

They're omitted so that we don't use workflow outputs as inputs to declarations, for example.

Having said that, I noticed that's still a bit too strict - #2860 - so we'll need to revisit this again soon

Reply:
To answer your second question, in the graph where we call a subworkflow, outputs will be regular output ports on a WorkflowCall node, rather than `GraphOutputDeclarationNode`s on the inner Node so I think this won't affect us there